src/blockchain/alt_block_tree.cpp

```diff
@@ -75,105 +77,200 @@ void commitPayloadsIds(BlockIndex<AltBlock>& index,
+                 "block %s already contains payloads",
+                 index.toPrettyString());
+
+  VBK_ASSERT_MSG(!index.hasFlags(BLOCK_CONNECTED),
+                 "state corruption: block %s is connected",
+                 index.toPrettyString());
+  VBK_ASSERT_MSG(!index.hasFlags(BLOCK_APPLIED),
+                 "state corruption: block %s is applied",
                  index.toPrettyString());
 
-  // prev block must exist
   VBK_ASSERT_MSG(index.pprev,
-                 "It is not allowed to add payloads to bootstrap block");
+                 "Adding payloads to a bootstrap block is not allowed");
+
+  // FIXME: this check belongs to popData
+  VBK_ASSERT_MSG(!hasDuplicateIds(payloads.context),
+                 "block %s must not contain duplicate context VBK blocks",
+                 index.toPrettyString());
+  VBK_ASSERT_MSG(!hasDuplicateIds(payloads.vtbs),
+                 "block %s must not contain duplicate VBKs",
+                 index.toPrettyString());
+  VBK_ASSERT_MSG(!hasDuplicateIds(payloads.atvs),
+                 "block %s must not contain duplicate ATVs",
+                 index.toPrettyString());
+
+  // add payload ids to the block, update the payload index
+  commitPayloadsIds<VbkBlock>(index, payloads.context, storage_);
+  commitPayloadsIds<VTB>(index, payloads.vtbs, storage_);
+  commitPayloadsIds<ATV>(index, payloads.atvs, storage_);
+
+  // save payloads on disk
+  storage_.savePayloads(payloads);
+
+  // we successfully added this block payloads
+  index.setFlag(BLOCK_HAS_PAYLOADS);
+}
+
+bool AltTree::connectBlock(index_t& index, ValidationState& state) {
+  VBK_ASSERT_MSG(index.hasFlags(BLOCK_HAS_PAYLOADS),
+                 "block %s must have payloads added",
+                 index.toPrettyString());
+  VBK_ASSERT_MSG(!index.hasFlags(BLOCK_CONNECTED),
+                 "block %s is already connected",
+                 index.toPrettyString());
+  VBK_ASSERT_MSG(!index.hasFlags(BLOCK_APPLIED),
+                 "state corruption: block %s is applied",
+                 index.toPrettyString());
 
   if (getParams().isStrictAddPayloadsOrderingEnabled()) {
-    // when StrictAddPayloadsOrdering is enabled, we do not allow to execute
-    // AddPayloads to applied chain
-    VBK_ASSERT_MSG(index.pprev->hasFlags(BLOCK_HAS_PAYLOADS),
-                   "pprev does not contain BLOCK_HAS_PAYLOADS %s",
+    VBK_ASSERT_MSG(index.pprev->hasFlags(BLOCK_CONNECTED),
+                   "the previous block of block %s must be connected",
+                   index.toPrettyString());
+    VBK_ASSERT_MSG(index.allDescendantsUnconnected(),
+                   "a descendant of block %s is connected",
+                   index.toPrettyString());
+  }
+
+  // partial stateful validation
+  // FIXME: eventually we want to perform full stateful validation here,
+  // effectively find out whether setState(index) will be successful
+
+  // BUG: applyBlock will clear BLOCK_FAILED_POP flag since it does not check
```

Comment:
could you elaborate?

Reply:
PopStateMachine.applyBlock() will happily attempt to apply a BLOCK_FAILED_POP block and clear the flag if it is successful, however applyBlock() doesn't perform complete validation of blocks. AFAIK, we aren't using this feature or "feature" at this moment, but eventually we need to either get rid of it or fix it.